File: medium/046_Permutations/permutations.py
class Solution(object):
    def permute_internal(self, a, start, result):
        if start == len(a):
            result.append(a[:])
            return

        for i in xrange(start, len(a)):
            a[i], a[start] = a[start], a[i]
            self.permute_internal(a, start+1, result)
            a[i], a[start] = a[start], a[i]

    def permute(self, a):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(a) == 0:
            return []

        result = []
        stack = []
        self.permute_internal(a, 0, result)
        return result

def test_case(a):
    result = solution.permute(a)
    print "Permutation for {}, cnt {}\n{}".format(a, len(result), result)

solution = Solution()
test_case([1])
test_case([1, 2, 3])
test_case([1, 2, 3, 4])

File: medium/046_Permutations/permutations.go
package main

import "fmt"

func makeCopy(src []int) []int {
	dst := make([]int, len(src))
	copy(dst, src)
	return dst
}

func permuteInternal(nums []int, start int, ans [][]int) [][]int {
	if start == len(nums) {
		ans = append(ans, makeCopy(nums))
	}
	for i := start; i < len(nums); i++ {
		nums[i], nums[start] = nums[start], nums[i]
		ans = permuteInternal(nums, start+1, ans)
		nums[i], nums[start] = nums[start], nums[i]
	}
	return ans
}

func permute(nums []int) [][]int {
	ans := make([][]int, 0)
	ans = permuteInternal(nums, 0, ans)
	return ans
}

func testPermute(nums []int) {
	ans := permute(nums)
	fmt.Printf("nums %v, ans %v\n", nums, ans)
}

func main() {
	testPermute([]int{1})
	testPermute([]int{1, 2})
	testPermute([]int{1, 2, 3})
}
